### PR TITLE
Prevent registration of an e-mail as username if it's already associated with existing user

### LINKED ISSUE #15024: E-mail accounts validation during the registration process
### Describe the bug

During the registration proccess an user is able to register herself with the same username as the email ID of another existing user. We should prevent that from happening.

## Workaround

- Enable "Email as username" flag or disable "Login with email" in the login settings

### Version

>= 19.0.3

### Expected behavior

- Not allowing the registration of an e-mail as username if it's already associated with an existing user.

### Actual behavior

- Allowing the registration of e-mail as username with the same e-mail ID of an existing user.

### How to Reproduce?

1. Register a user with username `test` having `test@test.com` as e-mail
2. Register another user with username `test@test.com`

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Vividious**
This PR tries to solve the issue: https://github.com/keycloak/keycloak/issues/15024

Resolves #15024

### CODE REVIEW

**Comment by mabartos** [comment]
@Vividious Thanks for the PR! 

However, I think it'd be good to provide some sort of flexibility for admins. I'd also suggest using the User Profile validator, as it abstracts all necessary validation,s even for the registration form. In your case, the problem is only solved for REST API.

I've created a draft PR for it (#15122 ).

File: services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java [hunk 1]
```diff
@@ -134,6 +134,12 @@ public Response createUser(final UserRepresentation rep) {
         if (session.users().getUserByUsername(realm, username) != null) {
             return ErrorResponse.exists("User exists with same username");
         }
+
+        // Check if username is already used as email by other user
+        if (session.users().getUserByEmail(realm, username) != null){
```
Comment (stianst) [inline services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:139]: This doesn't guarantee there isn't another user with this email address as the user could be registered just after this check, but prior to the new user being created.

**Review by stianst — CHANGES_REQUESTED** [review]
We had this one come up a while back, which we fixed by always looking up by email address first if the lookup has an '@' in it, which basically mean sure someone could register a user with a username same as another users email, but they won't be able to use it for login.

I disagree that this is a security issue/hardening issue for the above reasons.

File: services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java [hunk 2]
```diff
@@ -134,6 +134,12 @@ public Response createUser(final UserRepresentation rep) {
...
+            return ErrorResponse.exists("User name is already associated with an existing user");
+        }
+
```
[the remark this answers is not in the document]
Comment (stianst) [reply services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:142]: As long as self-service registration is open this makes no difference from an enumeration perspective.

[the remark this answers is not in the document]
Comment (stianst) [reply services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:142]: We already have the equivalent of what you are suggesting with disabling username, which is use email as username option or something like that.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Please see https://github.com/keycloak/keycloak/issues/15024#issuecomment-1293177416

File: services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java (hunk 1 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:139]: I second that. I'm in favour of closing this PR and looking for a proper solution in the original issue.

Comment (Vividious) [reply services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:139]: Aye, agree.
As it's one of my first PRs in OSS project, there is a question I need to ask. 
Should I actually create a PR draft before creating a real one?
Thanks!

**Pull request closed without merging**